forge/src/main/java/me/lucko/luckperms/forge/messaging/PluginMessageMessenger.java

```diff
@@ -0,0 +1,98 @@
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.luckperms.forge.messaging;
+
+import com.google.common.collect.Iterables;
+import io.netty.buffer.Unpooled;
+import me.lucko.luckperms.common.plugin.scheduler.SchedulerTask;
+import me.lucko.luckperms.forge.LPForgeBootstrap;
+import me.lucko.luckperms.forge.LPForgePlugin;
+import net.luckperms.api.messenger.IncomingMessageConsumer;
+import net.luckperms.api.messenger.Messenger;
+import net.luckperms.api.messenger.message.OutgoingMessage;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.protocol.Packet;
+import net.minecraft.network.protocol.game.ClientboundCustomPayloadPacket;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.players.PlayerList;
+import net.minecraftforge.network.NetworkRegistry;
+import net.minecraftforge.network.event.EventNetworkChannel;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class PluginMessageMessenger implements Messenger {
+    private static final ResourceLocation CHANNEL = new ResourceLocation(LPForgeBootstrap.ID, "update");
+
+    private final LPForgePlugin plugin;
+    private final IncomingMessageConsumer consumer;
+    private EventNetworkChannel channel;
+
+    public PluginMessageMessenger(LPForgePlugin plugin, IncomingMessageConsumer consumer) {
+        this.plugin = plugin;
+        this.consumer = consumer;
+    }
+
+    public void init() {
+        this.channel = NetworkRegistry.newEventChannel(CHANNEL, () -> "1", predicate -> true, predicate -> true);
+        this.channel.addListener(event -> {
+            String message = event.getPayload().readUtf();
+            this.consumer.consumeIncomingMessageAsString(message);
+            event.getSource().get().setPacketHandled(true);
+        });
+    }
+
+    @Override
+    public void sendOutgoingMessage(@NonNull OutgoingMessage outgoingMessage) {
+        AtomicReference<SchedulerTask> taskRef = new AtomicReference<>();
+        SchedulerTask task = this.plugin.getBootstrap().getScheduler().asyncRepeating(() -> {
+            ServerPlayer player = this.plugin.getBootstrap().getServer()
+                    .map(MinecraftServer::getPlayerList)
+                    .map(PlayerList::getPlayers)
+                    .map(players -> Iterables.getFirst(players, null))
+                    .orElse(null);
+            if (player == null) {
+                return;
+            }
+
+            FriendlyByteBuf byteBuf = new FriendlyByteBuf(Unpooled.buffer());
```

Comment:
Recently discovered that the utf8 format used by FriendlyByteBuf isn't quite the same as what LP expects.

If we rebase, can switch to use the newly added `AbstractPluginMessageMessenger` - copy the change made to Fabric: https://github.com/LuckPerms/LuckPerms/commit/c35958b88ec05022270461d2b8c522c8d991a008#diff-022aacd2913f3d175268a7ede8f4176df3b1e0739c96885d3599116a7595a2c7